tools/stats/stats.cpp

```diff
@@ -0,0 +1,195 @@
+#elif defined(_WIN32)
+  return FILE_ATTRIBUTE_DIRECTORY & GetFileAttributes(path);
+#else
+  return false;
+#endif
+}
+
+void AddInputDir(const char* path) {
+#ifdef __linux__
+  nftw(path, AddInputFile, 20, FTW_PHYS);
+#endif  // __linux__
+}
+
+void PrintUsage(char* argv0) {
+  printf(
+      R"(%s - Collect statistics from one or more SPIR-V binary file(s).
+
+USAGE: %s [options] [<filepaths>] [<dirpaths>]
+
+The SPIR-V binaries read from the given <filepaths> combined with the files
+found in the given <dirpaths> and their subdirectories.
+
+<dirpaths> feature is currently only implmeneted in Linux.
+
+Options:
+  -h, --help                       Print this help.
+)",
+      argv0, argv0);
+}
+
+void DiagnosticsMessageHandler(spv_message_level_t level, const char*,
+                               const spv_position_t& position,
+                               const char* message) {
+  switch (level) {
+    case SPV_MSG_FATAL:
+    case SPV_MSG_INTERNAL_ERROR:
+    case SPV_MSG_ERROR:
+      std::cerr << "error: " << position.index << ": " << message
+                << std::endl;
+      break;
+    case SPV_MSG_WARNING:
+      std::cout << "warning: " << position.index << ": " << message
+                << std::endl;
+      break;
+    case SPV_MSG_INFO:
+      std::cout << "info: " << position.index << ": " << message << std::endl;
+      break;
+    default:
+      break;
+  }
+}
+
+}  // namespace
+
+int main(int argc, char** argv) {
+  bool continue_processing = true;
+  int return_code = 0;
+
+  std::vector<const char*> arg_paths;
+
+  for (int argi = 1; continue_processing && argi < argc; ++argi) {
+    const char* cur_arg = argv[argi];
+    if ('-' == cur_arg[0]) {
+      if (0 == strcmp(cur_arg, "--help") || 0 == strcmp(cur_arg, "-h")) {
+        PrintUsage(argv[0]);
+        continue_processing = false;
+        return_code = 0;
+      } else {
+        PrintUsage(argv[0]);
+        continue_processing = false;
+        return_code = 1;
+      }
+    } else {
+      arg_paths.push_back(cur_arg);
+    }
+  }
+
+  // Exit if command line parsing was not successful.
+  if (!continue_processing) {
+    return return_code;
+  }
+
+  for (const char* path : arg_paths) {
+    if (IsDir(path)) {
+#ifndef __linux__
+      std::cerr << "error: Directory traversal only implemented on Linux"
+                << std::endl;
+      return 1;
+#endif  // __linux__
+      AddInputDir(path);
+    } else {
+      AddInputFile(path);
+    }
+  }
+
+  std::cerr << "Processing " << g_input_files.size()
+            << " files..." << std::endl;
+
+  spvtools::SpirvTools tools(SPV_ENV_UNIVERSAL_1_1);
+  tools.SetMessageConsumer(DiagnosticsMessageHandler);
+
+  libspirv::SpirvStats stats;
+
+  for (size_t index = 0; index < g_input_files.size(); ++index) {
+    constexpr size_t kMilestonePeriod = 1000;
+    if (index % kMilestonePeriod == kMilestonePeriod - 1)
+      std::cerr << "Processed " << index + 1 << " files..." << std::endl;
```

Comment:
This is not an error. So why is it going to cerr?

Recommend prefixing this by the program name, in case this is used with a lot of other tools. Also, guard it with a "-v" for "verbose" option.